### DISCUSSION TITLE
Filter functionality - Groovy or Java Packages?

### DISCUSSION THREAD

**Discussion by DanielVeale-Shipt** [provectus/kafka-ui#3780] [Q&A] [answered]
Hey everyone, 

I was wondering to what extent I could use Groovy or Java scripts to enhance my filters.

For example, given filter for a message that has `message.Timestamp1` x-hours after `message.Timestamp2`. 

I've tried doing this in Groovy. Using `import groovy.time.*` or prefixing the functions with their package names. Neither seemed to work.

Could anyone help out? Are we able to import packaging or to which extent are packages already included?

**Comment by Haarolean** [comment]
@DanielVeale-Shipt hey, it should work in general. Could you provide en example(-s) you've tried?

**Reply by DanielVeale-Shipt** [reply, marked as answer]
I was attempting

```
import groovy.time.*

def startString = "2023-05-02T16:44:55.362871439Z"
def endString = "2023-05-03T16:44:55.362871439Z"

def start = Date.parse("yyyy-MM-dd'T'HH:mm:ss",startString)
def end = Date.parse("yyyy-MM-dd'T'HH:mm:ss",endString)

def timeAllowed = new Duration(0,2,0,0,0)

TimeDuration duration = TimeCategory.minus(end, start)

duration.toMilliseconds() > timeAllowed.toMilliseconds()
```

^ This was never showing results despite always being true. 

However, I've successfully managed to another script to work using Java and its packages, by using 

`import java.text.SimpleDateFormat;  
  import java.util.Date;`

**Reply by DanielVeale-Shipt** [reply]
Hopefully this will speed up anyone else's R&D, thanks @Haarolean!

**Reply by Haarolean** [reply]
@DanielVeale-Shipt thank you

**Reply by mrittlinger** [reply]
@DanielVeale-Shipt Do you mind sharing the java example you were able to get to work?

**Reply by DanielVeale-Shipt** [reply]
@mrittlinger Sure thing, it was something like - 

```
import java.text.SimpleDateFormat;  
import java.util.Date;  

String sDate0 = value.time0; 
String sDate1=value.time1; 

Date date0=new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss").parse(sDate0);  
Date date1=new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss").parse(sDate1);  

long differenceInTime = date1.getTime() - date0.getTime();
long differenceInMinutes= (differenceInTime/ (1000 * 60));

differenceInMinutes< 1;
```
